RISE-V2G-Shared/src/main/java/com/v2gclarity/risev2g/shared/misc/V2GImplementationFactory.java

```diff
@@ -0,0 +1,73 @@
+/*******************************************************************************
+ * The MIT License (MIT)
+ *
+ * Copyright 2017 Dr.-Ing. Marc Mültin (V2G Clarity)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *******************************************************************************/
+package com.v2gclarity.risev2g.shared.misc;
+
+import java.lang.reflect.Constructor;
+import java.util.Arrays;
+
+import com.v2gclarity.risev2g.shared.utils.MiscUtils;
+
+/**
+ * This class serves as the base for implementation factory
+ * classes used in the SE/EV projects
+ * It will look up and instantiate a class based on a
+ * configuration property
+ */
+abstract public class V2GImplementationFactory {
+
+	/**
+	 * Builds an object instance from the configuration properties
+	 * The configuration should hold the class of the instance that
+	 * will be built.
+	 * @param propertyName Name of the property the contains the fully qualified class name
+	 * @param cls Target class of the build instance
+	 * @param params Optional arguments to the constructor
+	 * @return
+	 */
+	protected static <T> T buildFromProperties(String propertyName, Class<T> cls, Object...params) {
+		try {
+			String className = MiscUtils.getV2gEntityConfig().getProperty(propertyName);
+			if (className == null) {
+				return null;
+			}
+			
+			Class<?> clazz = Class.forName(className);
+			
+			Class<?>[] paramClasses = Arrays.stream(params)
```

Comment:
Could you please explain in more detail (using comments in the code) what exactly happens in the lines 57 to 59? I know that the idea is to instantiate a class given the property name,  but it's not totally clear to me what happens here with the .map and the .toArray.

Reply:
I admin this was a bit tricky.
Because the backend classes (BackendInterface, DummyACEVSEController, etc) are usually instantiated with V2GCommunicationSessionSECC as a parameter, I had to look up the correct constructor for the instantiation.

The clazz.getConstructor() needs and array with the classes of the parameters. I used the stream api here to convert the method parameters (params) which is an array of objects to an array of their classes.

But reviewing this part, I'm not quite sure this is the best approach for the builder. I tried to solve the problem without changing the already existing interfaces.

But if we had originally instead of:
```java
setBackendInterface(new DummyBackendInterface(this));
```
something like this:
```java
IBackendInterface backend = new DummyBackendInterface(this);
backend.setCommunicationSession(this);
setBackendInterface(backend);
```

then the instantiation would be as easy as `Class.forName(className).newInstance();`

What I see now is that there is a hidden requirement for the implementation classes: they need to provide a `constructor(V2GCommunicationSessionSECC commSessionContext)`.

So adding a 'setCommunicationSession()` method to  `IBackendInterface`, `IEVSEController` and `IEVController` would make things more transparent for the implementer.

What do you think about this?

Reply:
I inspected the code for the interfaces further and it turns out that we don't really need the V2GCommunicationSessionSECC for instantiating the objects for the interfaces IBackendInterface, IEVSEController, IACEVSEController, and IDCEVSEController.

The only point where we need the V2GCommunicationSessionSECC (that always comes as the parameter 'commSessionContext') is the method getSAScheduleList for the IBackendInterface. So that method needs to have an additional parameter 

`V2GCommunicationSessionSECC commSessionContext`

The only function calling getSAScheduleList is processIncomingMessage of the state WaitForChargeParameterDiscoveryReq. And there, we can just use 'getCommSessionContext()' which is provided via the superclass State when calling getSAScheduleList.

So in the end, this works fine:

`setACEvseController(new DummyACEVSEController());
  setDCEvseController(new DummyDCEVSEController());
  setBackendInterface(new DummyBackendInterface());`

And this should make your implementation easier I guess, right? 
Would the instantiation also be faster if we don't hand over this big commSessionContext variable? Not sure if that is has a major effect (it's just a pointer to a variable, right?)

On the DummyEVController, the commSessionContext is used several times. The thing is: I think you could avoid using the commSessionContext as a parameter for the constructor as well, but maybe there are some use cases where EVCC providers might want to make use of the session variable for things I didn't have in mind yet. It might be better to keep it there.

So, in case we leave the commSessionContext with all interfaces, your approach to extend the interfaces to have a setter for the commSessionContext that must be implemented is a valid solution I'd say. 

Let me also know what you think about whether or not we should keep the commSessionContext variable for the SECC interfaces.

Reply:
It's no secret that we're working on and SE implementation based on RISE-V2G. There I found the commSessionContext useful, because it can give us more insight about the current charge process. I've explained this in more detail in issue #31 

So I would keep a call to 'setCommSessionContext()' at the moment unless some other mechanism is provided. It is just a reference anyway, so does not take up much ram neither CPU.

Reply:
OK then, I'll be waiting for your updated code.

Reply:
I've pushed the updated code.

So now all the implementation classes have a default (parameterless) constructor, but the respective interfaces makes sure that they will contains a `setCommSessionContext` method.

Reply:
Looking good, thanks for the contribution.